database/src/ops/block.rs

```diff
@@ -1,89 +1,422 @@
 //! Blocks.
 
 //---------------------------------------------------------------------------------------------------- Import
+use monero_serai::transaction::Timelock;
 
-//---------------------------------------------------------------------------------------------------- Free Functions
+use cuprate_types::VerifiedBlockInformation;
+
+use crate::{
+    database::{DatabaseRo, DatabaseRw},
+    env::EnvInner,
+    error::RuntimeError,
+    tables::{
+        BlockBlobs, BlockHeights, BlockInfoV1s, BlockInfoV2s, BlockInfoV3s, KeyImages, NumOutputs,
+        Outputs, PrunableHashes, PrunableTxBlobs, PrunedTxBlobs, RctOutputs, Tables, TablesMut,
+        TxHeights, TxIds, TxUnlockTime,
+    },
+    transaction::{TxRo, TxRw},
+    types::{
+        BlockHash, BlockHeight, BlockInfoLatest, BlockInfoV1, BlockInfoV2, BlockInfoV3, KeyImage,
+        Output, PreRctOutputId, RctOutput,
+    },
+};
+
+//---------------------------------------------------------------------------------------------------- Private free functions
+// Some utility functions/types used internally in the public functions.
+
+//---------------------------------------------------------------------------------------------------- Free functions
 /// TODO
-pub fn add_block() {
-    todo!()
+///
+/// # Errors
+/// TODO
+#[inline]
+#[allow(clippy::too_many_lines)]
+pub fn add_block<'env, Ro, Rw, Env>(
+    env: &Env,
+    tx_rw: &Rw,
+    block: &VerifiedBlockInformation,
+) -> Result<(), RuntimeError>
+where
+    Ro: TxRo<'env>,
+    Rw: TxRw<'env>,
+    Env: EnvInner<'env, Ro, Rw>,
+{
+    add_block_inner(&mut env.open_tables_mut(tx_rw)?, block)
 }
 
 /// TODO
-pub fn add_block_data() {
-    todo!()
+///
+/// # Errors
+/// TODO
+#[inline]
+pub fn add_blocks<'env, Ro, Rw, Env>(
+    env: &Env,
+    tx_rw: &Rw,
+    blocks: &[VerifiedBlockInformation],
+) -> Result<(), RuntimeError>
+where
+    Ro: TxRo<'env>,
+    Rw: TxRw<'env>,
+    Env: EnvInner<'env, Ro, Rw>,
+{
+    let mut tables = env.open_tables_mut(tx_rw)?;
+    for block in blocks {
+        add_block_inner(&mut tables, block)?;
+    }
+    Ok(())
+}
```

Comment:
4. Now, our public "frontend" functions just open up the tables and pass it to `add_block_inner()` - and it is still efficient since we only open them once, even in the bulk case